runbooks/sectigo-root-ca-rotation-2025/linux.sh

```diff
@@ -0,0 +1,181 @@
+}
+
+record_pre_restart_state() {
+  echo "Recording log offsets before restart..."
+  : > "$OFFSETS_FILE"
+  for f in $LOG_FILES; do
+    if sudo test -f "$f"; then
+      size="$(sudo stat -c%s "$f" 2>/dev/null || echo 0)"
+      echo "$f:$size" >> "$OFFSETS_FILE"
+    else
+      echo "$f:-1" >> "$OFFSETS_FILE"
+    fi
+  done
+  PRE_TS_UNIX="$(date +%s)"
+  PRE_TS_READABLE="$(date -u '+%Y-%m-%d %H:%M:%S UTC')"
+  echo "Pre-restart timestamp: $PRE_TS_READABLE (epoch: $PRE_TS_UNIX)"
+}
+
+restart_agent() {
+  echo "Restarting the DataDog Agent..."
+  if ! sudo service datadog-agent restart; then
+    error_exit "Error: Failed to restart the DataDog agent."
+  fi
+  echo "Waiting 30 seconds for the DataDog Agent to restart..."
+  sleep 30
+}
+
+test_connectivity_since_restart() {
+  echo "=== Connectivity test (since this restart) ==="
+  local pattern='CERTIFICATE_VERIFY_FAILED|certificate verify failed|ssl[[:space:][:punct:]]*error'
+
+  # Scan only new bytes appended to each log since restart
+  while IFS= read -r line; do
+    log_file="${line%%:*}"
+    pre_size="${line##*:}"
+    if sudo test -f "$log_file"; then
+      if [ "$pre_size" -ge 0 ]; then
+        sudo tail -c "+$((pre_size+1))" "$log_file" 2>/dev/null | grep -qiE "$pattern" \
+          && error_exit "Detected SSL/cert verification failure in $(basename "$log_file") since restart."
+      else
+        sudo tail -n 500 "$log_file" 2>/dev/null | grep -qiE "$pattern" \
+          && error_exit "Detected SSL/cert verification failure in new $(basename "$log_file") since restart."
+      fi
+    fi
+  done < "$OFFSETS_FILE"
+
+  # Journald: only entries since PRE_TS
+  if command -v journalctl >/dev/null 2>&1; then
+    # Prefer epoch form; fall back to a readable UTC timestamp if needed
+    local SINCE_ARG="@${PRE_TS_UNIX}"
+    if ! sudo journalctl --since "$SINCE_ARG" -n 0 &>/dev/null; then
+      SINCE_ARG="$PRE_TS_READABLE"
+    fi
+    sudo journalctl -u datadog-agent --since "$SINCE_ARG" --no-pager | grep -qiE "$pattern" \
+      && error_exit "Detected SSL/cert verification failure in journald since restart."
+  fi
+
+  # Confirm API key
+  if sudo /etc/init.d/datadog-agent info 2>/dev/null | grep -q "API Key is valid"; then
+    echo "API key validation: OK"
+  else
+    echo "Warning: Could not confirm 'API Key is valid' from agent info." >&2
+  fi
+
+  echo "Connectivity test passed: no certificate verification errors since restart."
+}
+
+restore_environment() {
+  rm -f "$OFFSETS_FILE" || true
+  if [ "$TEMP_INSTALLED" -eq 1 ]; then
+    echo "Restoring environment: Removing temporarily installed curl."
+    if ! eval "$REMOVE_CMD"; then
+      echo "Warning: Failed to remove curl. Please remove it manually if necessary." >&2
+      echo "Please contact support for further help." >&2
+    else
+      echo "Temporarily installed curl has been removed."
+    fi
+  fi
+}
+
+main() {
+  check_downloader
+  ensure_target_directory
+  download_certificate
```

Comment:
should we test that the downloaded certificate works here via `wget --ca-certificate` or `curl --cert` before restarting the agent? it could indicate a failure earlier in the update process here